Parse/SMTLIB2.cpp

```diff
@@ -2040,32 +1851,59 @@ bool SMTLIB2::parseAsUserDefinedSymbol(const std::string& id,LExpr* exp,bool isS
   unsigned numTypeArgs = type->numTypeArguments();
   unsigned arity = symbol->arity();
 
-  static Stack<TermList> args;
-  args.reset();
-
-  Substitution subst;
-  for (unsigned i = 0; i < arity; i++) {
+  TermStack termArgs;
+  MatchingUtils::MapBinderAndApplicator subst;
+  for (unsigned i = numTypeArgs; i < arity; i++) {
     if (_results.isEmpty() || _results.top().isSeparator()) {
       complainAboutArgShortageOrWrongSorts("user defined symbol",exp);
     }
     TermList arg;
     auto argSort = _results.pop().asTerm(arg);
-    TermList sort;
-    if (i < numTypeArgs) {
-      // if the term has a parametric type, the parameters
-      // are checked and saved to instantiate the concrete
-      // types of later (non-type) arguments
-      subst.bind(type->quantifiedVar(i).var(), arg);
-      sort = type->arg(i);
+    // Sort matching does not work well, but our sorts
+    // are not dependent anyways, so just check equals
+    if (isSort) {
+      if (type->arg(i)!=argSort) {
+        complainAboutArgShortageOrWrongSorts("user defined symbol",exp);
+      }
     } else {
-      sort = SubstHelper::apply(type->arg(i), subst);
+      if (!MatchingUtils::matchTerms(type->arg(i), argSort, subst)) {
+        complainAboutArgShortageOrWrongSorts("user defined symbol",exp);
+      }
     }
+    termArgs.push(arg);
+  }
 
-    if (sort != argSort) {
-      complainAboutArgShortageOrWrongSorts("user defined symbol",exp);
+  if (_todo.isNonEmpty()) {
+    auto [op,exp] = _todo.top();
+    if (op == PO_AS_END) {
+      _todo.pop();
+      ASS(!isSort);
+      TermList resSort;
+      auto sort = _results.pop().asTerm(resSort);
+      ASS_EQ(sort, AtomicSort::superSort());
+      if (!MatchingUtils::matchTerms(type->result(), resSort, subst)) {
+        complainAboutArgShortageOrWrongSorts("user defined symbol",exp);
+      }
+    }
+  }
+
+  TermStack args;
+  for (unsigned i = 0; i < numTypeArgs; i++) {
+    auto typeVar = type->quantifiedVar(i).var();
+    TermList typeVarS;
+    // We require terms to have unambiguous sorts, except for ctor
+    // terms in 'match' blocks which are disambiguated by the sort
+    // of the matched term. We collect type variable instantiations
+    // from the arguments recursively, which by assumption already have
+    // unambiguous sorts. If any variable remains free after this, it
+    // means the result sort contains some free variable, and the user
+    // must enclose it in an (as <term> <sort>) block.
+    if (!subst._map.find(typeVar, typeVarS)) {
+      USER_ERROR_EXPR("User defined term "+exp->toString()+" has ambiguous sort, use (as "+exp->toString()+" <sort>) block to disambiguate");
```

Comment:
Nice - does the `(as _ sort)` parsing code now check that the term could actually be of that sort? A more specific sort is fine: `nil :: list int`, but more general is not: `cons 0 nil :: list alpha`.